tests/HourlyObjectOrientedTestWithPyomoExpression.py

```diff
@@ -0,0 +1,276 @@
+  drivers = np.zeros(life+1, dtype=object)
+  yearDfs = dfSet[0].groupby([dfSet[0].index.year])
+  oldList = []
+  newList = []
+  countOld = 0
+  countNew = 0
+  count = 0
+  for year, yearDf in yearDfs:
+    countOld = countNew
+    oldList.append(countOld)
+    if count != 0:
+      countNew = len(yearDf) + countOld
+      newList.append(countNew)
+      alphas[count] = dfSet[1][alpha][oldList[count]:newList[count]].mean()
+      drivers[count] = dfSet[1][driver][oldList[count]:newList[count]].mean()
+    else:
+      newList.append(countNew)
+      alphas[count] = 0
+      drivers[count] = 0
+    count += 1
+    if count == life+1:
+      break
+  # construct annual summary cashflows
+  cf.computeYearlyCashflow(alphas, drivers)
+  return cf
+
+def createRecurringHourly(dfSet, comp, driver, alpha):
+  """
+    Constructs recurring cashflow with one value per hour
+    @ In, dfSet, tuple, includes pandas.Dataframe, dict of inputs, and pyomo concrete model loaded
+    sto run
+    @ In, comp, CashFlow.Component, component this cf will belong to
+    @ In, driver, string, variable name in df to take driver from
+    @ In, alpha, string, variable name in df to take alpha from
+    @ Out, comps, dict, dict mapping names to CashFlow component objects
+  """
+  life = comp.getLifetime()
+  cf = CashFlows.Recurring()
+  cfFarams = {'name': 'RecursHourly',
+               'X': 1,
+               'mult_target': None,
+               'inflation': False}
+  cf.setParams(cfFarams)
+  cf.initParams(life, pyomoVar=dfSet[2])
+  yearDfs = dfSet[0].groupby([dfSet[0].index.year])
+  countOld = 0
+  countNew = 0
+  for year, yearDf in yearDfs:
+    countOld = countNew
+    countNew = len(yearDf) + countOld
+    y = year - 2018
+    if y > life:
+      break
+    cf.computeIntrayearCashflow(y, dfSet[1][driver][countOld:countNew], dfSet[1][alpha][countOld:countNew])
+  return cf
+
+
+
+if __name__ == '__main__':
+  # load multiyear data
+  ## TODO use analytic data! this is data from a not-propr-ietary report, but not analytic.
+  targets = ['A', 'B', 'C', 'D', 'Year', 'Time']
+  indices = ['RAVEN_sample_ID']
+  print('Loading data ...')
+  full_df = pd.read_csv('aux_file/hourly.csv',
+                        index_col=indices,
+                        usecols=targets+indices) #,
+                        #nrows=300000)
+  # just the first sample
+  df = full_df.loc[0]
+  years = pd.to_datetime(df['Year'].values + 2019, format='%Y')
+  hours = pd.to_timedelta(df['Time'].values, unit='H')
+  datetime = years + hours
+  df.index = datetime
+  df = df.sort_index()[['A', 'B', 'C', 'D']]
+
+  x = df.to_numpy()
+  A = np.zeros((x.shape[0]))
+  B = np.zeros((x.shape[0]))
+  C = np.zeros((x.shape[0]))
+  D = np.zeros((x.shape[0]))
+  for i in range(x.shape[0]):
+    A[i] = x[i][0]
+    B[i] = x[i][1]
+    C[i] = x[i][2]
+    D[i] = x[i][3]
+  dictDf = {}
+  dictDf['A'] = A
+  dictDf['B'] = B
+  dictDf['C'] = C
+  dictDf['D'] = D
+  dfSet = (df, dictDf)
```

Comment:
we need more clarity, perhaps from comments or variable naming, to explain what data is coming from the CSV and what data is pyomo vardata. Where possible, we should eliminate the interaction with the CSV.

Reply:
Thank you for leaving the comments above! Knowing your time is extremely limited, I appreciate the guidance and assistance!